lib/spack/spack/container/__init__.py

```diff
@@ -0,0 +1,73 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+"""Package that provides functions and classes to
+generate container recipes from a Spack environment
+"""
+import warnings
+
+import spack.schema.env as env
+import spack.util.spack_yaml as syaml
+from .writers import recipe
+
+__all__ = ['validate', 'recipe']
+
+
+def validate(configuration_file):
+    """Validate a Spack environment YAML file that is being used to generate a
+    recipe for a container.
+
+    Since a few attributes of the configuration must have specific values for
+    the container recipe, this function returns a sanitized copy of the
+    configuration in the input file. If any modification is needed, a warning
+    will be issued.
+
+    Args:
+        configuration_file (str): path to the Spack environment YAML file
+
+    Returns:
+        A sanitized copy of the configuration stored in the input file
+    """
+    import jsonschema
+    with open(configuration_file) as f:
+        config = syaml.load(f)
+
+    # The "container" subsection is mandatory
+    msg = '"container" subsection missing in "{0}"'
+    assert 'container' in config['spack'], msg.format(configuration_file)
+
+    # Remove attributes that are not needed / allowed in the
+    # container recipe
+    for subsection in ('cdash', 'gitlab_ci', 'modules'):
+        if subsection in config['spack']:
+            msg = ('the subsection "{0}" in "{1}" is not used when generating'
+                   ' container recipes and will be discarded')
+            warnings.warn(msg.format(subsection, configuration_file))
+            config['spack'].pop(subsection)
+
+    # Set the default value of the concretization strategy to "together" and
+    # warn if the user explicitly set another value
+    config['spack'].setdefault('concretization', 'together')
```

Comment:
This should probably make use of the `config_dict` functionality in `environment.py` that abstracts away the name of the top-level key in an environment.

Reply:
Thanks, I didn't know that function. Out of curiosity (I've been meaning to ask this for a while, but always forget): where do we use the `env` attribute as a substitute for `spack` in a `spack.yaml`?

Reply:
Done

Reply:
I believe the documentation says either is allowed. We initially had `env`, and Todd changed it to `spack` but wanted to maintain backwards compatibility IIRC.